### PR TITLE
Study about konnectors isolation solutions

### PR DESCRIPTION
**Pull request by doubleface**
To much detail, not enough. Remarks are welcome!

### CODE REVIEW

File: docs/konnectors.md [hunk 1]
```diff
@@ -208,6 +208,114 @@ single thread or application. But NaCl is [no longer
 maintained](https://bugs.chromium.org/p/chromium/issues/detail?id=239656#c160)
 and ZeroVM has some severe limitations, so, it won't be used.
 
+## Konnector isolation study
+
+The short list of tools which will be tested to isolate connectors is Rkt and NsJail which on paper better fullfill our needs.
+
+### NsJail
+
+NsJail is a lightweight process isolation tool, making use of Linux namespaces and seccomp-bpf syscall filters. It is not a container tool like docker. Its features are quiet extensive regarding isolation. The [README](https://github.com/google/nsjail) gives the full list of options available. Although available in the google github, it is not an official google tool.
```
Comment (nono) [inline docs/konnectors.md]: s/quiet/quite/

s/options available/available options/

File: docs/konnectors.md [hunk 2]
```diff
@@ -208,6 +208,114 @@ single thread or application. But NaCl is [no longer
...
+
+NsJail is:
+
+- easy to install : juste a make away with standard build tools
```
Comment (nono) [inline docs/konnectors.md]: s/juste/just/

File: docs/konnectors.md [hunk 3]
```diff
@@ -208,6 +208,114 @@ single thread or application. But NaCl is [no longer
...
+- Is still actively maintained
+
+### Rkt
+
+Rkt is very similar to docker. It can even directly run docker images from the docker registry, which gives us a lot of existing images to use, even if we want to be able to use other languages than node. For example, we could have also a container dedicated to weboob, another container could use phantomjs or casper and without forcing self-hosted users to do complicated installation procedures.
+
+Rkt is :
+- easy to install : debian, rpm package available, archlinux community package : https://github.com/coreos/rkt/releases
+- has network isolation like docker
+- offers CPU, memory limitation, seccomp isolation (but the set of rules to use is out of my understanding)
```
Comment (nono) [inline docs/konnectors.md]: Is it possible to customize the set of seccomp rules?

File: docs/konnectors.md [hunk 4]
```diff
@@ -208,6 +208,114 @@ single thread or application. But NaCl is [no longer
...
+- is well [documented](https://coreos.com/rkt/docs/latest/), complete man pages, but not as well known as docker, then there is not a lot of things to find outside the official documentation.
+- can use docker image directly or can convert them to one runnable aci file with one simple cli command (rkt export)
+- is in active developpement but relatively stable regarding core features.
+- container images can be easily signed and the signature is checked by default when running a container.
+
+I managed to run a nodejs container with just the following commands :
+
+    rkt run --insecure-options=image --interactive docker://node:slim --name nodeslim -- -v
```
Comment (nono) [inline docs/konnectors.md]: Can a non-root user launch `rkt run`?

File: docs/konnectors.md [hunk 5]
```diff
@@ -208,6 +208,114 @@ single thread or application. But NaCl is [no longer
...
+    rkt list   # to get the container uuid
+    rkt export --app=nodeslim <uuid> nodeslim.aci
+    rkt run --insecure-options=image --interactive nodeslim.aci -- -v  # to run node -v in the new container
+    
+Note: the --insecure-options param is to avoid the check of the image signature to ease the demonstration
+
+### Choice
+
+The best choice would be Rkt for it's ease of use (which is good for contribution) and wide range of isolation features + access to the big docker echosystem without beeing a burden for the host administrator.
```
Comment (nono) [inline docs/konnectors.md]: s/echosystem/ecosystem/

File: docs/konnectors.md [hunk 6]
```diff
@@ -208,6 +208,114 @@ single thread or application. But NaCl is [no longer
...
+
+#### Image creation
+
+To create an ACI file image, you just need to run a docker image one time :
+    
+    rkt run  --uuid-file-save=$PWD/uuid --insecure-options=image --interactive docker://node:slim --name nodeslim -- -v
+    rkt export --app=nodeslim `cat uuid` nodeslim.aci && rm uuid
+
+
+The node:slim image weights 84M at this time. The node:alpine image also exists and is way lighter (19M) but I had problems with DNS with this, which could be solved with more time.
```
Comment (nono) [inline docs/konnectors.md]: Alpine is not a good idea. It uses a libc that is not the glibc, and it has some nasty side effects: it introduces some bugs that are difficult to track.

cc: @aeris for more on this subject

File: docs/konnectors.md [hunk 7]
```diff
@@ -208,6 +208,114 @@ single thread or application. But NaCl is [no longer
...
+    rkt run --net=host --uuid-file-save=$PWD/uuid --volume data,kind=host,source=$PWD/container_dir --insecure-options=image nodeslim.aci --cpu=100m --memory=128M --name rktnode --mount volume=data,target=/usr/src/app --exec node -- /usr/src/app/$1 $2 &
+    # the container will handle itself the communication with the stack
+    sleep 60
+    rkt stop --force --uuid-file=uuid
+    rkt rm --uuid-file=uuid
+    rm -rf ./container_dir
+
+This script can be run like this :
+
+    ./rkt.sh mynewkonnector.js COZY_STACK_CREDENTIALS
```
Comment (nono) [inline docs/konnectors.md]: The credentials must be passed in env variables, not on the command-line. Else, they appear on `ps aux` and can be easily stolen by another user on the server.

File: docs/konnectors.md [hunk 8]
```diff
@@ -208,6 +208,114 @@ single thread or application. But NaCl is [no longer
...
+    rm -rf ./container_dir
+
+This script can be run like this :
+
+    ./rkt.sh mynewkonnector.js
+
+If the mynewkonnector.js file is available in the container_dir_template directory.
+
+Pros : no breach regarding localhost network, more secure
+Cons: a file communication protocol with the stack need to be defined (any other idea ?), it may be more difficult to migrate the existing connectors.
```
Comment (nono) [inline docs/konnectors.md]: I'm not sure to understand this part. The cozy-syack has a REST API for files, and the konnectors will use it the same way they will use the REST API for data.

File: docs/konnectors.md (hunk 7 above)
Comment (nono) [inline docs/konnectors.md]: We should also give the URL of the cozy instance to the konnector.

File: docs/konnectors.md [hunk 9]
```diff
@@ -208,6 +208,114 @@ single thread or application. But NaCl is [no longer
...
+
+    ./rkt.sh mynewkonnector.js COZY_STACK_CREDENTIALS
+
+If the mynewkonnector.js file is available in the container_dir_template directory.
+
+Pros: the container can directly the stack to make its own updates, like before-> less work on migrating konnectors
+Cons : must forbid acces to port 5984 and 6060 + SMTP server
+
+For both solutions, the limitation of time, CPU and memory will avoid most DOS attacks (to my knowledge). For memory use, I still don't see a way to prevent the excessive use of swap from the container.
+To prevent the connectors from listening to each other, they should be run one by one in their own container and not at the same time.
```
Comment (nono) [inline docs/konnectors.md]: Does it mean that we can't parallelize the konnectors by running several containers at the same time? If yes, it's a very serious inconvenient.

File: docs/konnectors.md (hunk 4 above)
Comment (doubleface) [inline docs/konnectors.md]: Yes, if they belong to the rkt group. I will add the information in the document

File: docs/konnectors.md (hunk 6 above)
Comment (doubleface) [inline docs/konnectors.md]: Ok, I will add this information about alpine

File: docs/konnectors.md (hunk 8 above)
Comment (doubleface) [inline docs/konnectors.md]: Ok, I was just trying to propose a solution to more completely isolate the konnectors. If the konnectors must use the REST-API, then the second solution is the only one. I will remove the first solution.

File: docs/konnectors.md (hunk 7 above)
Comment (doubleface) [inline docs/konnectors.md]: Ok, will update the script
@doubleface

File: docs/konnectors.md (hunk 9 above)
Comment (doubleface) [inline docs/konnectors.md]: No, we can absolutely run many containers at the same time. But in that case, they could spy other konnectors to get some credentials, except if the other connectors use https.

**Pull request merged**